libsignal-protocol/src/messages/signal_message.rs

```diff
@@ -119,4 +119,33 @@ impl From<SignalMessage> for CiphertextMessage {
     }
 }
 
+impl Serializable for SignalMessage {
+    fn deserialize(ctx: &Context, data: &[u8]) -> Result<Self, failure::Error> {
+        unsafe {
+            let mut raw = ptr::null_mut();
+
+            let res = sys::signal_message_deserialize(
+                &mut raw,
+                data.as_ptr(),
+                data.len(),
+                ctx.raw(),
+            );
+            if res != 0 {
+                return Err(failure::err_msg("Unable to deserialize buffer"));
+            }
+            // safety: the successful invocation of signal_message_deserialize
+            // tells us this is actually a pointer to a `signal_message`
+            let raw = Raw::copied_from(raw as *mut sys::signal_message);
+            Ok(SignalMessage {
+                raw,
+                _ctx: Rc::clone(&ctx.0),
+            })
+        }
+    }
+
+    fn serialize(&self) -> Result<Buffer, failure::Error> {
+        unimplemented!()
```

Comment:
I don't think this will ever be implemented as it is not supported by the underlying library. 

I would say that serialization is typically done as follows:

1. Encrypt message with `SessionCipher::encrypt(&self, message: &[u8]) -> Result<CiphertextMessage, Error>`
2. Serialize the returned `CiphertextMessage`
3. Send serialized message together with stuff like destination and message type (pre-key message, message, sender key...) to server.

Deserialization:

1. Fetch message envelope from server.
2. Check said envelope for type of message.
3. Depending on type, invoke `PreKeySignalMessage::deserialize` or `SignalMessage::deserialize`.
4. Depending on type, invoke either `SessionCipher::decrypt_pre_key_message` or `SessionCipher::decrypt_message`

This is why `CiphertextMessage` only implements `serialize` and `PreKeySignalMessage` and `SignalMessage` both implement `deserialize`.

It could be argued that the Rust API should operate at a higher level, but I probably think it is better to stick to the same semantics as the C library.

Reply:
What about situations where you want to use example messages as test fixtures? Under those circumstances it would be beneficial to implement both `serialize()` (for saving to disk in the first place) and `deserialize()` (for reading from disk during test setup)... But I'm guessing if it's not implemented by the C library our hands are tied, aren't they?

Reply:
Could you provide a more concrete example of what this would look like and why it would be something we would want?

Reply:
Some ramblings:

Regarding serialize: 
`libsignal-protocol-rs` doesn't provide an API to create `PreKeySignalMessage` or `SignalMessage` explicitly, so what would trigger the need to convert a `CiphertextMessage` to a `PreKeySignalMessage` or a `SignalMessage` before serializing it?

Regarding deserialize: 
If it wasn't for the fact that the server provides the type in the message envelope I think deserialization of `CiphertextMessage` would make more sense. To do it I think we would have to try to deserialize to `PreKeySignalMessage` or `SignalMessage` and see what succeeds and then upcast to `CiphertextMessage`. This would then have to be converted to a `PreKeySignalMessage` or `SignalMessage` before being useful. 

_If we were to implement it (which I am not suggesting) it could maybe make sense to return an `enum` which would either contain `PreKeySignalMessage` or `SignalMessage`. Maybe `CiphertextMessage` would be that enum?_